src/foundry_dev_tools/resources/dataset.py

```diff
@@ -0,0 +1,620 @@
+        if self.transaction is not None:
+            self._context.catalog.api_abort_transaction(self.rid, self.transaction["rid"])
+            self._transaction = None
+
+    def upload_files(
+        self,
+        path_file_dict: dict[api_types.PathInDataset, Path],
+        max_workers: int | None = None,
+        **kwargs,
+    ) -> api_types.Transaction:
+        """Uploads multiple local files to a foundry dataset.
+
+        Args:
+            max_workers: Set number of threads for upload
+            path_file_dict: A dictionary which maps the path in the dataset -> local file path
+            **kwargs: get passed to :py:meth:`foundry_dev_tools.resources.dataset.Dataset.transaction_context`
+        """
+        with self.transaction_context(**kwargs):
+            self._context.data_proxy.upload_dataset_files(
+                dataset_rid=self.rid,
+                transaction_rid=self.transaction["rid"],
+                path_file_dict=path_file_dict,
+                max_workers=max_workers,
+            )
+            return self.transaction
+
+    def upload_file(
+        self,
+        file_path: Path,
+        path_in_foundry_dataset: api_types.PathInDataset,
+        **kwargs,
+    ) -> api_types.Transaction:
+        """Upload a file to the dataset.
+
+        Args:
+            file_path: local file path to upload
+            path_in_foundry_dataset: file path inside the foundry dataset
+            transaction_type: if this dataset does not have an open transaction,
+                opens a transaction with the specified type.
+            start_transaction: if true, a transaction will be created if no transaction is currently open
+            finish_transaction: if true, the transaction will be closed (committed or aborted) after the upload.
+            **kwargs: get passed to :py:meth:`foundry_dev_tools.resources.dataset.Dataset.transaction_context`
+        """
+        with self.transaction_context(**kwargs):
+            self._context.data_proxy.upload_dataset_file(
+                dataset_rid=self.rid,
+                transaction_rid=self.transaction["rid"],
+                path=file_path,
+                path_in_foundry_dataset=path_in_foundry_dataset,
+            )
+            return self.transaction
+
+    def upload_folder(self, folder_path: Path, **kwargs) -> api_types.Transaction:
+        """Uploads all files contained in the folder to the dataset.
+
+        The default transaction type is UPDATE.
+
+        Args:
+            folder_path: the folder to upload
+            **kwargs: passed to :py:meth:`foundry_dev_tools.resources.dataset.Dataset.upload_files`
+
+        Returns:
+            the transaction used for uploading the files
+        """
+        kwargs.setdefault("transaction_type", api_types.FoundryTransaction.UPDATE)
+        return self.upload_files({str(f.relative_to(folder_path)): f for f in folder_path.rglob("*")}, **kwargs)
+
+    def save_model(
```

Comment:
I think we should leave it up to the user to pickle/dill/serialize the model and not just use pickle. 
Maybe we just delete this function? I also don't really like it on the CachedFoundryClient.

Reply:
Remove from dataset class and think about new design.